### PR TITLE
[GEOS-10837] geopackage output fails when java.io.tmpdir on network share

### PR DESCRIPTION
**Pull request by davidblasby**
[GEOS-10837](https://osgeo-org.atlassian.net/browse/GEOS-10837)

This allows the user to set a property (i.e. `-Dgeopackage.getfeatureoutput.tempdir=/path/to/temp`) that uses an alternative temp dir for the geopackage.  For people with the standard java temp dir being a mounted network drive, they should use this (pointing to a local drive).

GeoPKG (SQL Lite) has a lot of problems locking when writing to cloud (or network-mounted) file systems.

NOTE: I moved the test cases into the "proper" package (to match the class-under-test `GeoPackageGetFeatureOutputFormat`).
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [n/a] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [n/a] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/extension/geopkg-output/src/main/java/org/geoserver/geopkg/wfs/GeoPackageGetFeatureOutputFormat.java [hunk 1]
```diff
@@ -71,14 +78,42 @@ protected String getExtension(FeatureCollectionResponse response) {
...
+     * @return
+     */
+    File createTempFile(String prefix, String suffix) throws IOException {
+        String customTempDir = GeoServerExtensions.getProperty(CUSTOM_TEMP_DIR_PROPERTY);
+        if (!StringUtils.hasText(customTempDir)) {
+            return File.createTempFile("geopkg", ".tmp.gpkg");
+        }
+        File tempDir = new File(customTempDir);
+        if (!tempDir.exists() || !tempDir.isDirectory()) {
+            throw new IOException(
```
Comment (jodygarnett) [inline src/extension/geopkg-output/src/main/java/org/geoserver/geopkg/wfs/GeoPackageGetFeatureOutputFormat.java:101]:
```suggestion
            throw new FileNotFoundException(
```

Will need to adjust import also

File: src/extension/geopkg-output/src/main/java/org/geoserver/geopkg/wfs/GeoPackageGetFeatureOutputFormat.java [hunk 2]
```diff
@@ -71,14 +78,42 @@ protected String getExtension(FeatureCollectionResponse response) {
...
+     *
+     * <p>If the CUSTOM_TEMP_DIR_PROPERTY property is set, then use that as the temp directory.
+     * Otherwise (unset), use the standard File.createTempFile() temp directory.
+     *
+     * <p>If the CUSTOM_TEMP_DIR_PROPERTY directory doesn't exist (or isnt a directory), then an
+     * exception is thrown
+     *
+     * @param prefix
+     * @param suffix
+     * @return
```
Comment (jodygarnett) [inline src/extension/geopkg-output/src/main/java/org/geoserver/geopkg/wfs/GeoPackageGetFeatureOutputFormat.java:92]:
```suggestion
     * @param prefix filename prefix, must be three characters or more
     * @param suffix filename suffix, or if null {@code null} suffix {@code ".tmp") used
     * @return path to newly created temporary file
```

**Comment by mprins** [comment]
I'd rather use `java.io.tmpdir` or  `CATALINA_TMPDIR` instead of adding a property that only fixes this for one format.

**Comment by jodygarnett** [comment]
> I'd rather use `java.io.tmpdir` or `CATALINA_TMPDIR` instead of adding a property that only fixes this for one format.

This is explicitly to avoid a ``java.io.tmpdir`` setting which as been set to a network drive. Apparently that is an acceptable approach when running some kind of docker image with hardly any local disk.

This approach is somewhat similar to how WPS has a resource storage directory.

**Comment by mprins** [comment]
OK, I knew it had to be something exotic 😺

**Comment by jodygarnett** [comment]
@davidblasby I am taking this out of draft. The change you have done looks clean; and I do understand why it is required. I am going to add a note for the documentation (which is my only feedback).

**Review by jodygarnett — CHANGES_REQUESTED** [review]
Only useful feedback is to add a note to the docs which I will do now before merge and backport. Thank you for the improvement.

**Pull request merged**